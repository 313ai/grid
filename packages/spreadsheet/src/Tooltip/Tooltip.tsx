import React from "react";
import { TooltipProps } from "@rowsncolumns/grid";
import { INVALID_COLOR, HYPERLINK_COLOR } from "../constants";
import { Box, Link } from "@chakra-ui/core";
import { CellConfig } from "../Spreadsheet";
import { DATATYPE } from "../types";

export interface TipProps extends TooltipProps, CellConfig {
  valid?: boolean;
  content?: string;
  onMouseEnter?: () => void;
  onMouseLeave?: () => void;
  variant?: TooltipVariant;
  position?: TooltipPosition;
}

export type TooltipPosition = "right" | "left" | "top" | "bottom";
export type TooltipVariant = "error" | "info";

const Tooltip: React.FC<TipProps> = ({
  x = 0,
  y = 0,
  width = 0,
  position = "right",
  scrollLeft = 0,
  scrollTop = 0,
  height = 0,
  valid,
  variant = "info",
  content,
  datatype,
  hyperlink,
  text,
  onMouseEnter,
  onMouseLeave,
}) => {
  const title = valid === false ? "Invalid:" : "";
  const variantColor = variant === "error" ? INVALID_COLOR : "transparent";

  const posX = position === "right" ? x + width - scrollLeft : x - scrollLeft;
  const posY = position === "bottom" ? y + height - scrollTop : y - scrollTop;
  return (
    <div
      style={{
        position: "absolute",
        left: 0,
        top: 0,
        transform: `translate(${posX}px, ${posY}px)`,
        maxWidth: 200,
        minWidth: 160,
        background: "white",
        boxShadow: "0 4px 8px 3px rgba(60,64,67,.15)",
        padding: 12,
        borderRadius: 4,
        fontSize: 13,
        borderLeft: `4px ${variantColor} solid`,
        backfaceVisibility: "hidden",
      }}
      onMouseLeave={onMouseLeave}
      onMouseEnter={onMouseEnter}
    >
      {datatype === DATATYPE.Hyperlink ? (
        <HyperLink title={text} url={hyperlink} variantColor={variantColor} />
      ) : (
        <Alert title={title} content={content} variantColor={variantColor} />
      )}
    </div>
  );
};

interface TooltipContentProps {
  title?: React.ReactText;
  url?: string;
  content?: string;
  variantColor?: string;
}

const HyperLink: React.FC<TooltipContentProps> = ({ title, url }) => {
  return (
    <>
      <Box mb={1} fontWeight="bold">
        <Link color={HYPERLINK_COLOR} href={url} target="_blank">
          {title}
        </Link>
      </Box>
      <Box color="#666">{url}</Box>
    </>
  );
};

const Alert: React.FC<TooltipContentProps> = ({
  title,
  content,
  variantColor,
}) => {
  return (
    <>
      {title && (
        <Box pb={2} color={variantColor}>
          <strong>{title}</strong>
        </Box>
      )}
      {content}
    </>
  );
};

export default Tooltip;
