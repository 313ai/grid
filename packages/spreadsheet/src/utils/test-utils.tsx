import React from "react";
import { render } from "@testing-library/react";
import { Stage, Layer } from "react-konva";
import { ThemeProvider, theme } from "@chakra-ui/core";
import Konva from "konva";

const KonvaWrapper = ({ children }) => {
  return (
    <Stage>
      <ThemeWrapper>
        <Layer>{children}</Layer>
      </ThemeWrapper>
    </Stage>
  );
};

const konvaRenderer = (ui) => render(ui, { wrapper: KonvaWrapper });
const domRenderer = (ui) => render(ui, { wrapper: ThemeWrapper });

export * from "@testing-library/react";
export { konvaRenderer, domRenderer };

export const ThemeWrapper = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};
