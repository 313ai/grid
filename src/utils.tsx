import React from "react";
import { Line, Rect } from "react-konva";
import { ShapeConfig } from "konva/types/Shape";

interface BoxProps extends ShapeConfig {
  onCrosshairMouseDown?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseMove?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseUp?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

/**
 * Create a box with custom top/right/bottom/left colors and widths
 * @param param0
 */
export const createBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
}: BoxProps) => {
  const commonProps = {
    perfectDrawEnabled: false,
    shadowForStrokeEnabled: false,
    hitStrokeWidth: 0,
    listening: false,
    lineCap: "square",
  };
  const lines = [
    <Line
      points={[x, y, x + width, y]}
      stroke={strokeTopColor}
      strokeWidth={strokeTopWidth}
      key="top"
      {...commonProps}
    />,
    <Line
      points={[x + width, y, x + width, y + height]}
      stroke={strokeRightColor}
      strokeWidth={strokeRightWidth}
      key="right"
      {...commonProps}
    />,
    <Line
      points={[x + width, y + height, x, y + height]}
      stroke={strokeBottomColor}
      strokeWidth={strokeBottomWidth}
      key="bottom"
      {...commonProps}
    />,
    <Line
      points={[x, y + height, x, y]}
      stroke={strokeLeftColor}
      strokeWidth={strokeLeftWidth}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <Rect
          fill={fill}
          x={x}
          y={y}
          width={width}
          height={height}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};

export const FillHandle: React.FC<BoxProps> = ({
  x = 0,
  y = 0,
  stroke,
  onMouseDown,
}) => {
  if (x === 0 || y === 0) return null;
  const crossHairSize = 8;
  const crossHairBorderSize = 1;
  return (
    <div
      style={{
        position: "absolute",
        left: x - crossHairSize / 2,
        top: y - crossHairSize / 2,
        width: crossHairSize,
        height: crossHairSize,
        border: `${crossHairBorderSize}px white solid`,
        background: stroke,
        cursor: "crosshair",
      }}
      onMouseDown={onMouseDown}
    />
  );
};

export const createHTMLBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
  strokeStyle = "solid",
}: BoxProps) => {
  const commonProps = {};
  width = width - Math.floor(strokeWidth);
  y = y - Math.ceil(strokeWidth / 2);
  const lines = [
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: width,
        height: strokeTopWidth,
        borderWidth: 0,
        borderColor: strokeTopColor,
        borderTopWidth: strokeTopWidth,
        borderStyle: strokeStyle,
      }}
      key="top"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x + width,
        top: y,
        width: strokeRightWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeRightColor,
        borderRightWidth: strokeRightWidth,
        borderStyle: strokeStyle,
      }}
      key="right"
      {...commonProps}
    />,
    ,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y + height,
        width: width + strokeTopWidth,
        height: strokeBottomWidth,
        borderWidth: 0,
        borderColor: strokeBottomColor,
        borderBottomWidth: strokeBottomWidth,
        borderStyle: strokeStyle,
      }}
      key="bottom"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: strokeLeftWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeLeftColor,
        borderLeftWidth: strokeLeftWidth,
        borderStyle: strokeStyle,
      }}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <div
          style={{
            position: "absolute",
            top: y,
            left: x,
            height,
            width,
            backgroundColor: fill,
            userSelect: "none",
            pointerEvents: "none",
          }}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};
