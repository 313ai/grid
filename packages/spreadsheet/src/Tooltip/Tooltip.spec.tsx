import React, { useState, useRef } from "react";
import { domRenderer, cleanup, fireEvent, act } from "./../utils/test-utils";
import Tooltip from "./Tooltip";

describe("Tooltip", () => {
  afterEach(cleanup);
  it("renders select tooltip", () => {
    const renderGrid = () => domRenderer(<Tooltip  />);
    expect(renderGrid).not.toThrow();
  });
  it("matches snapshot", () => {
    const { asFragment } = domRenderer(<Tooltip />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders hyperlink", () => {
    const { asFragment } = domRenderer(<Tooltip datatype='hyperlink' result='Google' hyperlink='http://google.com' />);
    expect(asFragment()).toMatchSnapshot();
  });
});