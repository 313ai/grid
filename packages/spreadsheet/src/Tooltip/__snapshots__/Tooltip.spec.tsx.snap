// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px); max-width: 200px; min-width: 160px; background: white; box-shadow: 0 4px 8px 3px rgba(60,64,67,.15); padding: 12px; border-radius: 4px; font-size: 13px; border-left: 4px solid #1D72E8; user-select: none; word-break: break-word;"
  />
</DocumentFragment>
`;

exports[`Tooltip renders hyperlink 1`] = `
<DocumentFragment>
  <div
    style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px); max-width: 200px; min-width: 160px; background: white; box-shadow: 0 4px 8px 3px rgba(60,64,67,.15); padding: 12px; border-radius: 4px; font-size: 13px; border-left: 4px solid #1D72E8; user-select: none; word-break: break-word;"
  >
    <div
      class="css-1mg6gu6"
    >
      <a
        class="css-1enwuer"
        href="http://google.com"
        target="_blank"
      >
        Google
      </a>
    </div>
    <div
      class="css-2c7s1y"
    >
      http://google.com
    </div>
  </div>
</DocumentFragment>
`;
